### PR TITLE
feat(api): Explicit cmdline and jupyter entrypoints

### LINKED ISSUE #3546: Apiv2 Last Mile: Well-defined entrypoint for jupyter
Add an entrypoint usable from Jupyter (or interactive python shell) for running a protocol. There should be one for both protocol apis 1 and 2.

This should probably take the form of something that returns an instantiated protocol context that can have methods called on it. It should connect to the hardware controller using #3545 .

### LINKED ISSUE #3755: Apiv2: Add entrypoint for executing protocols from command line
## overview

In APIv2, you can't just run a python file like a script to run the protocol anymore - there's a lot of setup that's required. We need a script entrypoint that can run on the robot, like the simulator script but for actually executing a protocol.

### PR DESCRIPTION
**Pull request by sfoster1**
Create explicit entrypoints for commandline and jupyter execution of protocols, and refactor protocol parse and execution to pull more logic out of the rpc layer.

There's a new `opentrons_execute` entrypoint that would only be valid on a robot, and a `get_protocol_api` method that can be used from jupyter. `opentrons_execute` works regardless of protocol api version.

## Testing
- Install and make sure the entrypoints work (unless you install the buildroot system build, you won't have the `opentrons_execute` binding but you can still run it with `python -m opentrons.execute)`
- Run python and json protocols to check that the refactors didn't break anything

Closes #3755, closes #3546

### CODE REVIEW

File: api/src/opentrons/protocols/parse.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    assigns = [
+        obj for obj in parsed.body if isinstance(obj, ast.Assign)]
+    for obj in assigns:
+        # XXX This seems brittle and could probably do with
+        # - enough love that we can remove the type: ignores
+        # - some thought about exactly what types are allowed in metadata
+        if isinstance(obj.targets[0], ast.Name) \
+                and obj.targets[0].id == 'metadata' \
+                and isinstance(obj.value, ast.Dict):
+            keys = [k.s for k in obj.value.keys]  # type: ignore
```
Comment (IanLondon) [inline api/src/opentrons/protocols/parse.py]: this does not work if your `metadata` is nested - which it probably should be allowed to be?

```
    values = [v.s for v in obj.value.values]  # type: ignore
AttributeError: 'Dict' object has no attribute 's'
```

File: api/src/opentrons/protocols/parse.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        text=protocol_contents, filename=filename, contents=protocol_json)
+
+
+def _parse_python(
+        protocol_contents: str, filename: str = None) -> PythonProtocol:
+    """ Parse a protocol known or at least suspected to be python """
+    filename_checked = filename or '<protocol>'
+    parsed = ast.parse(protocol_contents,
+                       filename=filename_checked)
+    metadata = extract_metadata(parsed)
```
[the remark this answers is not in the document]
Comment (IanLondon) [reply api/src/opentrons/protocols/parse.py]: Convo on Slack: we'll figure out a general solution for parameters later, it seems crappy to require them in the .py, for PL we'll put stuff in `metadata` in a hacky PL-specific fashion

File: api/tests/opentrons/data/testosaur.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from opentrons import containers, instruments, robot
+from opentrons import instruments, containers, robot
```
Comment (Laura-Danielle) [inline api/tests/opentrons/data/testosaur.py]: was there a reason for this change? It seems a bit trivial.

File: api/tests/opentrons/legacy_protocols/test_execute_v1.py [hunk 4]
```diff
@@ -212,3 +213,10 @@ def mock_set_flow_rate(aspirate, dispense):
         (123, 102),
         (101, 102)
     ]
+
+
+def test_legacy_jsonprotocol_v1(get_json_protocol_fixture):
+    robot.reset()
+    protocol_data = get_json_protocol_fixture('1', 'simple', False)
+    protocol = parse(protocol_data, None)
```
Comment (Laura-Danielle) [inline api/tests/opentrons/legacy_protocols/test_execute_v1.py]: Do we want to verify the schema version that parse determines?

File: api/tests/opentrons/protocols/test_parse.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    assert check(prot_api1_no_meta_a) == expected[prot_api1_no_meta_a]
+    assert check(prot_api1_no_meta_b) == expected[prot_api1_no_meta_b]
+    assert check(prot_api1_no_meta_c) == expected[prot_api1_no_meta_c]
+    assert check(prot_api1_meta1) == expected[prot_api1_meta1]
+    assert check(prot_api1_meta2) == expected[prot_api1_meta2]
+    assert check(prot_api2_no_meta) == expected[prot_api2_no_meta]
+    assert check(prot_api2_meta1) == expected[prot_api2_meta1]
+    assert check(prot_api2_meta2) == expected[prot_api2_meta2]
+
+    # for protocol, expected_version in expected.items():
```
Comment (Laura-Danielle) [inline api/tests/opentrons/protocols/test_parse.py]: Can we get rid of this if we're not using it?

File: api/tests/opentrons/data/testosaur.py (hunk 3 above)
Comment (sfoster1) [reply api/tests/opentrons/data/testosaur.py]: It avoids #3949

File: api/tests/opentrons/legacy_protocols/test_execute_v1.py (hunk 4 above)
Comment (sfoster1) [reply api/tests/opentrons/legacy_protocols/test_execute_v1.py]: https://github.com/Opentrons/opentrons/pull/4032/files#diff-ac4f423a98d7e6968682166b7be0cadaR142 should do that

File: api/src/opentrons/protocol_api/execute.py [hunk 6]
```diff
@@ -195,28 +192,24 @@ def run_protocol(protocol_code: Any = None,
...
+    else:
+        protocol_version = get_protocol_schema_version(protocol.contents)
         if protocol_version > 3:
             raise RuntimeError(
                 f'JSON Protocol version {protocol_version} is not yet ' +
                 'supported in this version of the API')
 
-        validate_protocol(protocol_json)
-
         if protocol_version >= 3:
```
[the remark this answers is not in the document]
Comment (sfoster1) [reply api/src/opentrons/protocol_api/execute.py]: It will previously yell at you if it's greater than 3 so I think it's fine

**Comment by sfoster1** [comment]
Tested:

- get_protocol_api from jupyter
- opentrons_execute for apiv1 and apiv2 protocols and json on apiv1 and apiv2
- runapp protocol execution for apiv1 and apiv2 and json on apiv1 and apiv2
- opentrons_simulate on osx and windows for apiv1 and apiv2 protocols and json on apiv1 and apv2

File: api/docs/source/v2/apiv2index.rst [hunk 7]
```diff
@@ -14,6 +14,7 @@ Opentrons API Version 2
     new_complex_commands
     new_protocol_api
     new_examples
```
[the remark this answers is not in the document]
Comment (sfoster1) [reply api/docs/source/v2/apiv2index.rst]: #4092

**Pull request merged**